### PR TITLE
Redux+Saga

### PR DESCRIPTION
**Pull request by deman4ela**

### CODE REVIEW

File: src/Alert.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    return (
+      <div>
+       Unfortunately, {`${props.fetchedComicsError}`} error occurred. Please try again later.
+      </div>);
+  }
+
+  if (props.fetchedHeroesError) {
+    return (
+      <div>
+       Unfortunately, {`${props.fetchedHeroesError}`} error occurred. Please try again later.
```
Comment (alinalaktionova) [inline src/Alert.tsx]: the message is the same, you can simplify and create one if statement

File: src/redux/connects/connectHeroes.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import Heroes from '../../Components/Heroes';
+import { fetchHeroes } from '../actions';
+
+const mapStateToProps = (state: any) => ({
+  fetchedHeroesSuccess: state.heroes.fetchedHeroesSuccess,
+  fetchedHeroesError: state.heroes.fetchedHeroesError,
+  loaderWorksHeroes: state.heroes.loaderWorksHeroes
+});
+
+const mapDispatchToProps = (dispatch: any) => ({
```
Comment (alinalaktionova) [inline src/redux/connects/connectHeroes.tsx]: remove any

File: src/Components/ComicsOfHero.tsx [hunk 3]
```diff
@@ -5,33 +5,25 @@ import {
...
-      isLoading: true
-    };
-  }
 
   componentDidMount() {
-    const { match } = this.props;
-    getAllComicsOfHero(match.params.heroID).then((results) => {
-      this.setState({ comics: results, isLoading: false });
-    });
+    (() => this.props.fetchComics(this.props.match.params.heroID))();
```
Comment (alinalaktionova) [inline src/Components/ComicsOfHero.tsx]: why do you use anonymous func for this?

File: src/Components/Heroes.tsx [hunk 4]
```diff
@@ -8,28 +8,33 @@ import ProgressBar from './ProgressBar';
...
+      searchInput: ''
     };
     this.handleChange = this.handleChange.bind(this);
     this.handleSubmit = this.handleSubmit.bind(this);
   }
 
   componentDidMount() {
-    this.getHeroes();
+    let searchedHeroName = queryString.parse(this.props.location.search).query;
+    if (searchedHeroName === undefined) {
```
Comment (alinalaktionova) [inline src/Components/Heroes.tsx]: JUST searchedHeroName

File: src/Components/Heroes.tsx [hunk 5]
```diff
@@ -8,28 +8,33 @@ import ProgressBar from './ProgressBar';
...
+      searchedHeroName = '';
+    }
+    (() => this.props.fetchHeroes(searchedHeroName))();
```
Comment (alinalaktionova) [inline src/Components/Heroes.tsx]: again anonymous func

File: src/api/api.tsx [hunk 6]
```diff
@@ -30,3 +30,13 @@ export function getAllHeroes() {
 export function getAllComicsOfHero(heroID: any) {
   return getAPIResource(`characters/${heroID}/comics`);
 }
+
+export function getHeroes(searchedHeroName: any): any {
+  let result;
+  if (searchedHeroName === '') {
```
Comment (alinalaktionova) [inline src/api/api.tsx]: no need to compare with empty string

File: src/redux/actions.tsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { FETCH_HEROES, FETCH_HEROES_SUCCESS, FETCH_HEROES_ERROR, FETCH_COMICS, FETCH_COMICS_SUCCESS, FETCH_COMICS_ERROR, SHOW_LOADER_HEROES, HIDE_LOADER_HEROES, SHOW_LOADER_COMICS, HIDE_LOADER_COMICS }  from './types';
+
+export const fetchHeroes = (heroName: string) => ({ type: FETCH_HEROES, payload: heroName });
+export const fetchHeroesSuccess = (heroes: any) => ({ type: FETCH_HEROES_SUCCESS, payload: heroes });
```
Comment (alinalaktionova) [inline src/redux/actions.tsx]: remove any

File: src/redux/actions.tsx [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+export const fetchHeroesError = (err: any) => ({ type: FETCH_HEROES_ERROR, payload: err });
+
+export const showLoaderHeroes = () => ({ type: SHOW_LOADER_HEROES });
```
Comment (alinalaktionova) [inline src/redux/actions.tsx]: you can show loader in fetchHeroes action, it's not necessary to create separate action for this

File: src/redux/types.tsx [hunk 9]
```diff
@@ -0,0 +1,13 @@
+export const FETCH_HEROES = 'HERO/FETCH_HEROES';
```
Comment (alinalaktionova) [inline src/redux/types.tsx]: why HERO/ ?

Comment (alinalaktionova) [inline src/redux/types.tsx]: just FETCH_HEROES

File: src/redux/types.tsx [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+export const FETCH_HEROES_SUCCESS = 'HERO/FETCH_HEROES_SUCCESS';
+export const FETCH_HEROES_ERROR = 'HERO/FETCH_HEROES_ERROR';
+
+export const SHOW_LOADER_HEROES = 'HERO/SHOW_LOADER';
+export const HIDE_LOADER_HEROES = 'HERO/HIDE_LOADER';
+
+export const FETCH_COMICS = 'COMIC/FETCH_COMICS';
```
Comment (alinalaktionova) [inline src/redux/types.tsx]: the same

File: src/Components/Heroes.tsx [hunk 11]
```diff
@@ -44,37 +49,19 @@ class Heroes extends React.Component<any, any> {
...
-    } else {
-      getAllHeroes().then((results) => {
-        this.setState({ heroes: results, isLoading: false });
-      });
-    }
-  }
-
   render() {
-    const { heroes, isLoading } = this.state;
+    const { fetchedHeroesSuccess, fetchedHeroesError, loaderWorksHeroes } = this.props;
```
Comment (alinalaktionova) [inline src/Components/Heroes.tsx]: loaderWorksHeroes is bad naming, rename

**Review by alinalaktionova — CHANGES_REQUESTED** [review]

File: src/Components/ComicsOfHero.tsx (hunk 3 above)
Comment (deman4ela) [reply src/Components/ComicsOfHero.tsx]: I used an example from some source and did not change it. 
Now, it is changed and anonymous function is removed.

File: src/Components/Heroes.tsx (hunk 5 above)
Comment (deman4ela) [reply src/Components/Heroes.tsx]: Changed

File: src/Components/Heroes.tsx (hunk 4 above)
Comment (deman4ela) [reply src/Components/Heroes.tsx]: Changed to:
if (!searchedHeroName)

File: src/Components/Heroes.tsx (hunk 11 above)
Comment (deman4ela) [reply src/Components/Heroes.tsx]: Renamed to 'loaderForHeroes'

File: src/api/api.tsx (hunk 6 above)
Comment (deman4ela) [reply src/api/api.tsx]: Fixed.

File: src/redux/actions.tsx [hunk 12]
```diff
@@ -1,15 +1,13 @@
-import { FETCH_HEROES, FETCH_HEROES_SUCCESS, FETCH_HEROES_ERROR, FETCH_COMICS, FETCH_COMICS_SUCCESS, FETCH_COMICS_ERROR, SHOW_LOADER_HEROES, HIDE_LOADER_HEROES, SHOW_LOADER_COMICS, HIDE_LOADER_COMICS }  from './types';
+import { FETCH_HEROES, FETCH_HEROES_SUCCESS, FETCH_HEROES_ERROR, FETCH_COMICS, FETCH_COMICS_SUCCESS, FETCH_COMICS_ERROR, HIDE_LOADER_HEROES, HIDE_LOADER_COMICS }  from './types';
 
 export const fetchHeroes = (heroName: string) => ({ type: FETCH_HEROES, payload: heroName });
 export const fetchHeroesSuccess = (heroes: any) => ({ type: FETCH_HEROES_SUCCESS, payload: heroes });
 export const fetchHeroesError = (err: any) => ({ type: FETCH_HEROES_ERROR, payload: err });
 
-export const showLoaderHeroes = () => ({ type: SHOW_LOADER_HEROES });
 export const hideLoaderHeroes = () => ({ type: HIDE_LOADER_HEROES });
```
Comment (alinalaktionova) [inline src/redux/actions.tsx]: you can hide loader in success or error action and not to create separate one for this logic

**Review by alinalaktionova — CHANGES_REQUESTED** [review]

**Review by alinalaktionova — APPROVED** [review]

**Pull request merged**